lib/rpush/daemon/fcm/delivery.rb

```diff
@@ -0,0 +1,160 @@
+          else
+            fail Rpush::DeliveryError.new(response.code.to_i, @notification.id, Rpush::Daemon::HTTP_STATUS_CODES[response.code.to_i])
+          end
+        end
+
+        def ok
+          reflect(:fcm_delivered_to_recipient, @notification)
+          mark_delivered
+          log_info("#{@notification.id} sent to #{@notification.device_token}")
+        end
+
+        def bad_request(response)
+          fail Rpush::DeliveryError.new(400, @notification.id, "FCM failed to handle the JSON request. (#{parse_error(response)})")
+        end
+
+        def unauthorized
+          fail Rpush::DeliveryError.new(401, @notification.id, 'Unauthorized, Bearer token could not be validated.')
+        end
+
+        def sender_id_mismatch
+          fail Rpush::DeliveryError.new(403, @notification.id, 'The sender ID was mismatched. It seems the device token is wrong.')
+        end
+
+        def unregistered(response)
+          fail Rpush::DeliveryError.new(404, @notification.id, "Client was not registered for your app. (#{parse_error(response)})")
+        end
+
+        def too_many_requests
+          fail Rpush::DeliveryError.new(429, @notification.id, 'Slow down. Too many requests were sent!')
+        end
+
+        def internal_server_error(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM responded with an Internal Error. " + retry_message)
+        end
+
+        def bad_gateway(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM responded with a Bad Gateway Error. " + retry_message)
+        end
+
+        def service_unavailable(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM responded with an Service Unavailable Error. " + retry_message)
+        end
+
+        def other_5xx_error(response)
+          retry_delivery(@notification, response)
+          log_warn("FCM responded with a 5xx Error. " + retry_message)
+        end
+
+        def parse_error(response)
+          error = multi_json_load(response.body)['error']
+          puts "PARSED: #{error} #{error['message']}"
+          "#{error['status']}: #{error['message']}"
+        end
+
+        def deliver_after_header(response)
+          Rpush::Daemon::RetryHeaderParser.parse(response.header['retry-after'])
+        end
+
+        def retry_delivery(notification, response)
+          time = deliver_after_header(response)
+          if time
+            mark_retryable(notification, time)
+          else
+            mark_retryable_exponential(notification)
+          end
+        end
+
+        def retry_message
+          "Notification #{@notification.id} will be retried after #{@notification.deliver_after.strftime('%Y-%m-%d %H:%M:%S')} (retry #{@notification.retries})."
+        end
+
+        def obtain_access_token
+          authorizer = ::Google::Auth::ServiceAccountCredentials.make_creds(scope: SCOPE)
+          authorizer.fetch_access_token
```

Comment:
I could only find the bang-! version of this. https://googleapis.dev/ruby/googleauth/latest/Signet/OAuth2/Client.html#fetch_access_token!-instance_method. Is there an alias for it somewhere that I'm missing?

In general, I'm wondering if this is the most efficient implementation. As far as I understand, this will always request a new token, even if it hasn't expired yet.

[This article](https://blog.mmyoji.com/posts/2019/02-16-get-fcm-access-token-in-ruby/) mentions caching it. I don't think we need to use Google's cache stores but could use `Rails.cache` for it. There's also a [`.needs_access_token?`](https://googleapis.dev/ruby/googleauth/latest/Signet/OAuth2/Client.html#needs_access_token%3F-instance_method) helper. For this we'd need at least v1.1.0 of `googleauth` according to https://github.com/googleapis/google-auth-library-ruby/pull/347.

What do you think?

Some debug logs would probably help, because I don't recall how often `Delivery` is being instantiated off the top of my head.